### ISSUE TITLE
Function referring to itself

### ISSUE THREAD

**Issue by shinderohitt** [akash-akya/exscheme#1]
Akya!

testcase 
```elixir
  test "fact" do
    expr = "(begin (define fact (lambda (n) (if (< n 2) 1 (* n (fact (- n 1)))))) (fact 4))"
    {result, _env} = interprete(expr)
    assert result == 24
  end
```

Gives `[error] variable: fact not found in the env`.

Stumbled upon this same problem in my implementation, tried looking up how people are doing it, and came across this repo.

S, this is how Elixir behaves:

<img width="333" alt="Screenshot 2019-11-10 at 5 56 50 PM" src="https://user-images.githubusercontent.com/1011044/68544408-e9eada00-03e8-11ea-98e5-08d3d74b2336.png">

So _while_ "defining"  `fact`, we pass in global environment (which doesn't have reference to `fact`). So, inside `fact`'s body, `fact` is undefined. After `fact` is set in global context, global environment is modified, but it doesn't take effect in user defined function..

atm, I don't think there's a way to solve this problem. Can't really implement scheme in Elixir (or any immutable language). But I may be wrong :)

**Comment by akash-akya** [comment]
You are right, previous code was incorrect (there still maybe other issues) thanks for pointing it out. 

I pushed new changes. This seems to be the correct way of doing it.

Because of immutablity we can not keep env frame pointer in prcedure, as any set/define which needs to update frame will create new copy of env frame. Trick is to rather than keeping the reference to frame, keep an unique idetifier which points to frame. since the identifier will be fixed for a context/frame, any define/set don't have to update this identifer.
